File: server/gameserver/src/main/java/me/eddiep/ghost/gameserver/api/network/packets/ActionRequestPacket.java
package me.eddiep.ghost.gameserver.api.network.packets;

import me.eddiep.ghost.network.packet.Packet;
import me.eddiep.ghost.gameserver.api.network.TcpUdpClient;
import me.eddiep.ghost.gameserver.api.network.TcpUdpServer;

import java.io.IOException;

public class ActionRequestPacket extends Packet<TcpUdpServer, TcpUdpClient> {
    public ActionRequestPacket(TcpUdpClient client, byte[] data) {
        super(client, data);
    }

    @Override
    public void onHandlePacket(TcpUdpClient client)  throws IOException {
        int packetNumber = consume(4).asInt();
        if (packetNumber < client.getLastReadPacket()) {
            int dif = client.getLastReadPacket() - packetNumber;
            if (dif >= Integer.MAX_VALUE - 1000) {
                client.setLastReadPacket(packetNumber);
            } else return;
        } else {
            client.setLastReadPacket(packetNumber);
        }

        byte actionType = consume(1).asByte();
        float mouseX = consume(4).asFloat();
        float mouseY = consume(4).asFloat();
        //long time = consume(4).asLong();

        if (actionType == 0)
            client.getPlayer().moveTowards(mouseX, mouseY);
        else if (actionType >= 1)
            client.getPlayer().fireTowards(mouseX, mouseY, actionType);
        else
            System.err.println("[SERVER] Unknown action " + actionType + " ! (" + client.getIpAddress() + ")");
    }
}

File: client/core/src/com/boxtrotstudio/ghost/client/network/packets/ActionRequestPacket.kt
package com.boxtrotstudio.ghost.client.network.packets

import com.boxtrotstudio.ghost.client.network.Packet
import com.boxtrotstudio.ghost.client.network.PlayerClient

class ActionRequestPacket : Packet<PlayerClient>() {

    override fun write(vararg args : Any) {
        val action = args[0] as Byte
        val x = args[1] as Float
        val y = args[2] as Float

        client.sendCount++

        write(0x08.toByte())
        write(client.sendCount)
        write(action)
        write(x)
        write(y)
        endUdp()
    }
}
